inlang/source-code/sdk/src/loadProject.ts

```diff
@@ -448,3 +846,420 @@ export function createSubscribable<T>(signal: () => T): Subscribable<T> {
+				}
+			} while (messsageId === undefined)
+
+			// create a humanId based on a hash of the alias
+			loadedMessage.id = messsageId
+
+			const importedEnecoded = stringifyMessage(loadedMessage)
+
+			// add the message - this will trigger an async file creation in the backgound!
+			messagesQuery.create({ data: loadedMessage })
+			// we load a fresh version - lets delete dirty flag that got created by the create method
+			delete messageDirtyFlags[loadedMessage.id]
+			messageLoadHash[loadedMessage.id] = importedEnecoded
+		}
+	}
+	await releaseLock(fs as NodeishFilesystem, lockFilePath, "loadMessage", lockTime)
+
+	console.log("loadMessagesViaPlugin: " + loadedMessages.length + " Messages processed ")
+
+	isLoading = false
+
+	const executingScheduledMessages = sheduledLoadMessagesViaPlugin
+	if (executingScheduledMessages) {
+		// a load has been requested during the load - executed it
+
+		// reset sheduling to except scheduling again
+		sheduledLoadMessagesViaPlugin = undefined
+
+		// recall load unawaited to allow stack to pop
+		loadMessagesViaPlugin(fs, lockFilePath, messagesQuery, settingsValue, loadPlugin).then(
+			() => {
+				executingScheduledMessages[1]()
+			},
+			(e: Error) => {
+				executingScheduledMessages[2](e)
+			}
+		)
+	}
+}
+
+async function saveMessagesViaPlugin(
+	fs: NodeishFilesystemSubset,
+	lockFilePath: string,
+	messagesQuery: InlangProject["query"]["messages"],
+	settingsValue: ProjectSettings,
+	savePlugin: any
+): Promise<any> {
+	// queue next save if we have a save ongoing
+	if (isSaving) {
+		if (!sheduledSaveMessages) {
+			sheduledSaveMessages = createAwaitable()
+		}
+
+		return sheduledSaveMessages[0]
+	}
+
+	// set isSavingFlag
+	isSaving = true
+
+	currentSaveMessagesViaPlugin = (async function () {
+		const persistedMessageHashs = {} as { [messageId: string]: string }
+
+		// check if we have any dirty message - witho
+		if (Object.keys(messageDirtyFlags).length == 0) {
+			// nothing to save :-)
+			console.log("save was skiped - no messages marked as dirty...")
+			return
+		}
+
+		try {
+			const lockTime = await accquireFileLock(fs as NodeishFilesystem, lockFilePath, "saveMessage")
+
+			// since it may takes some time to accquire the lock we check if the save is required still (loadMessage could have happend in between)
+			if (Object.keys(messageDirtyFlags).length == 0) {
+				return
+			}
+
+			const currentMessages = messagesQuery.getAll()
+
+			const messagesToExport: Message[] = []
+			for (const message of currentMessages) {
+				if (messageDirtyFlags[message.id]) {
+					const importedEnecoded = stringifyMessage(message)
+					// TODO #1844 use hash instead of the whole object JSON to save memory...
+					persistedMessageHashs[message.id] = importedEnecoded
+				}
+
+				const fixedExportMessage = { ...message }
+				// TODO #1585 here we match using the id to support legacy load message plugins - after we introduced import / export methods we will use importedMessage.alias
```

Comment:
Would this code become redundant if `repo.fs.readFile()` would be "reactive" and take care of triggering only if a file, and thus message, changed?